### PR TITLE
Revise HMC to support general potential_fn

### PR DESCRIPTION
**Pull request by fehiepsi**
As part of #1816, this PR tries to revise HMC class to support general `potential_fn`. This feature is requested by some users in Pyro forum (e.g. to run HMC with custom or unnormalized log_prob) and GPyTorch folks.

The plan of this PR is to have a working version first, then we can discuss about what'd be needed to do with the interface later.

TODO:
- [x] Implements `initialize_model` function to create `intial_params`, `potential_fn`, and `transform_fn` (to generate `trace` of the current params)
- [x] Change the input of HMC/NUTS to `potential_fn` instead of `model`. And `initial_trace` is replaced by `initial_params`.
- [x] Think about a compatible layer for the current API. (need discussions about this when the above tasks are done).
- [x] Update tests if necessary (or maybe not)

### CODE REVIEW

**Comment by neerajprad** [comment]
@fehiepsi : As you correctly point out there are a few layers to this task. I would suggest refactoring as follows:
 1. Make the TraceKernel class take in dict instead of trace, any interconversion between trace/dict happens inside the MCMC class. - Straight refactor, no change to external interface, current tests work as expected.
 2. A new optional parameter called `potential_fn` is introduced. Current model kwarg remains but is deprecated. If model is provided, we call `initiallize_model` to get `potential_fn` and `init_params`.
 3. The above changes are still backwards compatible, except some deprecated arguments. Next we can have a light alternate API layer that doesn't subclass from `TracePosterior`.

These can all be separate PRs. I would be happy to chip in, if there is any part that you would like me to work on.

**Comment by fehiepsi** [comment]
Wow, your suggestions are really helpful! Thanks @neerajprad !

I would be great if you can code on your first point (refactoring TraceKernel class). :)

**Comment by neerajprad** [comment]
> I would be great if you can code on your first point (refactoring TraceKernel class)

Great, I'll get started on that!

**Comment by neerajprad** [comment]
@fehiepsi - could you confirm that there is no regression in total test times for `test_hmc.py` and `test_nuts.py`?

**Comment by fehiepsi** [comment]
Sure, I'll do it.

File: pyro/infer/mcmc/util.py [hunk 1]
```diff
@@ -208,3 +212,135 @@ def log_prob(self, model_trace):
...
+
+    dims = [frame.dim
+            for site in sites
+            for frame in site["cond_indep_stack"]
+            if frame.vectorized]
+    max_plate_nesting = -min(dims) if dims else 0
+    return max_plate_nesting
+
+
+def _transform_fn(transforms, params, invert=True):
```
Comment (neerajprad) [inline pyro/infer/mcmc/util.py]: Great to see this refactored out!

File: pyro/infer/mcmc/util.py [hunk 2]
```diff
@@ -208,3 +212,135 @@ def log_prob(self, model_trace):
...
+    return {k: transforms[k](v) if not invert else transforms[k].inv(v)
+            for k, v in params.items()}
+
+
+def _pe_maker(model, model_args, model_kwargs, trace_prob_evaluator, transforms):
+    def potential_energy(params):
+        params_constrained = {k: transforms[k].inv(v) for k, v in params.items()}
+        cond_model = poutine.condition(model, params_constrained)
```
Comment (neerajprad) [inline pyro/infer/mcmc/util.py]: Just a quick note here - `poutine.condition` will convert all stochastic sites into observed sites. I don't think that should really change anything at this point, but just in case you see any weird results, this is worth checking.

File: pyro/infer/mcmc/hmc.py [hunk 3]
```diff
@@ -30,6 +27,8 @@ class HMC(TraceKernel):
     Radford M. Neal
 
     :param model: Python callable containing Pyro primitives.
+    :param potential_fn: Python callable calculating potential energy with input
```
Comment (neerajprad) [inline pyro/infer/mcmc/hmc.py]: maybe also specify that the input parameters have real support.

File: pyro/infer/mcmc/util.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/infer/mcmc/util.py]: Hmm, I thought that I have deleted this function because it is not used elsewhere. ^^!

Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: I meant `initialize_model` and its contents. I think this method can be deleted.

**Comment by fehiepsi** [comment]
@neerajprad 

The following benchmark is for pre-this PR vs this PR

HMC:
+ logistic regression: 19.98s vs 20.52s
+ dirichlet categorical: 6.62s vs 6.37s
+ beta bernoulli: 8.09s vs 7.34s

NUTS:
+ gamma normal: 5.50s vs 5.23s
+ gamma beta: 13.90s vs 12.24s
+ gaussian mixture model: 26.54s vs 21.71s

Except HMC logistic regression, there is a gain with this PR. All diagnostic infos (acc rate, step size) are identical.

**Comment by neerajprad** [comment]
> Except HMC logistic regression, there is a gain with this PR. All diagnostic infos (acc rate, step size) are identical.

Sounds great! I generally look at the total test time. For individual tests, it is worth checking whether the chosen step size changed due to a different initial trace for instance. That can lead to different times, but that's acceptable.

File: pyro/infer/mcmc/util.py [hunk 4]
```diff
@@ -208,3 +212,131 @@ def log_prob(self, model_trace):
...
+
+def _get_init_params(model, model_args, model_kwargs, transforms, potential_fn, prototype_params,
+                     max_tries_initial_params=100):
+    params = prototype_params
+    for i in range(max_tries_initial_params):
+        potential_energy = potential_fn(params)
+        if not torch_isnan(potential_energy) and not torch_isinf(potential_energy):
+            return params
+        trace = poutine.trace(model).get_trace(*model_args, **model_kwargs)
+        samples = {name: trace.nodes[name]["value"].detach() for name in params}
```
Comment (neerajprad) [inline pyro/infer/mcmc/util.py]: Do we need to `.detach()` here since the only place where we set/unset `requires_grad` is the verlet integrator?

File: pyro/infer/mcmc/util.py [hunk 5]
```diff
@@ -208,3 +212,131 @@ def log_prob(self, model_trace):
...
+    model_trace = poutine.trace(model).get_trace(*model_args, **model_kwargs)
+    has_enumerable_sites = False
+    prototype_samples = {}
+    for name, node in model_trace.iter_stochastic_nodes():
+        if isinstance(node["fn"], _Subsample):
+            continue
+        if node["fn"].has_enumerate_support:
+            has_enumerable_sites = True
+            continue
+        prototype_samples[name] = node["value"].detach()
```
Comment (neerajprad) [inline pyro/infer/mcmc/util.py]: Same here, do we need `.detach()` here?

**Comment by neerajprad** [comment]
@fehiepsi - Since travis seems to be taking forever, I'm happy to merge this PR as is and do follow-up PRs if needed.

File: pyro/infer/mcmc/util.py (hunk 4 above)
Comment (fehiepsi) [reply pyro/infer/mcmc/util.py]: It seems that without it, there is an error happen. I'll check it against.

Comment (fehiepsi) [reply pyro/infer/mcmc/util.py]: Yes, it is required when our model has `pyro.param(...)` statements (so sample sites will be leaf variables, which make us unable to change `requires_grad` flag in velocity_verlet)

File: pyro/infer/mcmc/util.py [hunk 6]
```diff
@@ -208,3 +212,131 @@ def log_prob(self, model_trace):
...
+        optimized executable trace in the integrator.
+    :param dict jit_options: A dictionary contains optional arguments for
+        :func:`torch.jit.trace` function.
+    :param bool ignore_jit_warnings: Flag to ignore warnings from the JIT
+        tracer when ``jit_compile=True``. Default is False.
+    :returns: a tuple of (`initial_params`, `potential_fn`, `transforms`, `prototype_trace`)
+    """
+    # XXX `transforms` domains are sites' supports
+    if transforms is None:
+        automatic_transform_enabled = True
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply pyro/infer/mcmc/util.py]: I still can't find a good pattern here so I keep the current behaviour and would like to defer the change to another PR. The simplest change is to drop `transforms` arg. The better one (in my opinion), which covers a large number of usage cases, is to have `transforms={}` as default args and add default transforms for each latent site which does not appear in user-provided `transforms` arg. There we will need to check if codomain for each user-provided transform is `real` and raise an error when users provide wrong incompatible transforms.

[the remark this answers is not in the document]
Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: I think your proposed change would be the best.

File: pyro/infer/mcmc/util.py (hunk 4 above)
Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: I see. That because `pyro.param` also sets the `requires_grad` flag to True, I guess.

**Comment by fehiepsi** [comment]
@neerajprad We can resolve Memory error at https://github.com/pyro-ppl/pyro/issues/1730 with the strategy in your last PR (which is only yield params in workers):

```
class MCMC(TracePosterior):
    def __init__(...):
        self._model = kernel.model

    def _trace_wrap(self, z, *args, **kwargs):
        #if self.kernel.model is None:
        #    return z
        for name, transform in self._transforms.items():
            z[name] = transform.inv(z[name])
        z_trace = self._prototype_trace
        for name, value in z.items():
            z_trace.nodes[name]["value"] = value
        trace_poutine = poutine.trace(poutine.replay(self._model, z_trace))
        trace_poutine(*args, **kwargs)
        return trace_poutine.trace

    def _traces(self, *args, **kwargs):
        from pyro.infer.mcmc.util import initialize_model
        _, _, self._transforms, self._prototype_trace = initialize_model(self._model, args, kwargs)

        for sample in self.sampler._traces(*args, **kwargs):
            trace = self._trace_wrap(sample[0], *args, **kwargs)
            yield (trace, *sample[1:])
```
However this way will affect the performance because the main process will do a bunch of replaying job to get trace (at least this is the case for `baseball` example). How about we consume only samples (after transforming from params) instead of traces? And these samples will be populated when users call `posterior.exec_traces` if they want to do something with traces. At least, to get marginal posterior, we don't need these traces, only params and transforms are enough. This way also resolves https://github.com/pyro-ppl/pyro/issues/1787, where traces whole a bunch of big and non-used info such as `nodes['data']['fn'].mean`.

I guess you have thought about all of these issues with your big refactoring intention. I just want to share some of my observations and thoughts which might be helpful for you. :)

File: pyro/infer/mcmc/util.py (hunk 4 above)
Comment (fehiepsi) [reply pyro/infer/mcmc/util.py]: I think so too.

**Comment by neerajprad** [comment]
> How about we consume only samples (after transforming from params) instead of traces? And these samples will be populated when users call posterior.exec_traces

That's a good observation. I think the problem with the current interface is that MCMC inherits from `TracePosterior` and methods like `_traces` that return an iterator over traces. I think rather than catering to this interface at all, it might be better to build new posterior / predictive wrappers and address these issues there. Hopefully, we can then go and deprecate `TracePosterior` if the newly designed interface works well.

**Comment by jiamings** [comment]
Thanks for the implementation! But I am not sure if there is a bug in the HMC code, it seems that the potential_fn argument in __init__() is never used? The self.potential_fn is always overridden by the one derived from the model.
